### PR TITLE
Port torchvision models to CrypTen

### PR DESCRIPTION
**Pull request by knottb**
Summary: - Added `crypten.models` module that imports all torchvision models and modifies them to be `crypten.nn.Module`s

Differential Revision: D25033864

### CODE REVIEW

File: crypten/models/__init__.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    """
+    Returns a copy of an imported module so it can be modified
+    without modifying future imports of the given module
+    """
+    starting_modules = sys.modules.copy()
+
+    module_spec = importlib.util.find_spec(module_name)
+    module = importlib.util.module_from_spec(module_spec)
+    module_spec.loader.exec_module(module)
+    new_modules = set(sys.modules) - set(starting_modules)
```
Comment (gmuraru) [inline crypten/models/__init__.py:51]: NITPICK: Maybe use directly?
```
sys.modules.keys() - starting_modules.keys()?
```
^^ This should return a new set.
The ```keys``` method on a ```dict``` should give back a view for the keys and in this case there won't be created 2 new sets (one for ```sys.modules``` and one for ```starting_modules```)

File: crypten/models/__init__.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+    del module_spec
```
Comment (gmuraru) [inline crypten/models/__init__.py:53]: Q: Why this needs to be deleted? (wouldn't it get deleted when the function finishes?)

File: crypten/models/__init__.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        module.nn = cnn
+
+    # TODO: fix replacement in global `torch` module - perhaps use __torch_function__
+    if hasattr(module, "torch"):
+        module.torch.flatten = lambda x, *args: x.flatten(*args)
+        module.torch.transpose = lambda x, *args: x.transpose(*args)
+        # module.torch.cat = lambda *args, **kwargs: args[0].cat(*args, **kwargs)
+
+    if hasattr(module, "F"):
+        module.F = __functional_replacement()
```
Comment (gmuraru) [inline crypten/models/__init__.py:129]: Q: Can't ```__functional_replacement``` be used directly? (without creating an instance)

File: crypten/mpc/mpc.py [hunk 4]
```diff
@@ -753,7 +764,14 @@ def max_pool2d(self, kernel_size, padding=None, stride=None, return_indices=Fals
 
     @mode(Ptype.arithmetic)
     def _max_pool2d_backward(
-        self, indices, kernel_size, padding=None, stride=None, output_size=None
+        self,
+        indices,
+        kernel_size,
+        padding=None,
```
Comment (gmuraru) [inline crypten/mpc/mpc.py:770]: Q: Can't the padding be set here to 0 (by default), in that case, there is no need to check if the padding is ```None```

File: crypten/mpc/mpc.py [hunk 5]
```diff
@@ -769,9 +787,15 @@ def _max_pool2d_backward(
             stride = kernel_size
         if isinstance(stride, int):
             stride = stride, stride
-        assert isinstance(padding, tuple), "stride must be a int, tuple, or None"
+        assert isinstance(stride, tuple), "stride must be a int, tuple, or None"
         s0, s1 = stride
 
+        # Setup dilation
+        if isinstance(stride, int):
```
Comment (gmuraru) [inline crypten/mpc/mpc.py:794]: BUG: Should be ```dilation``` and at the assert it should be
```"dilation must be a int or tuple``` (the default value is set to 1)

**Comment by knottb** [comment]
This has landed. Closing.
